src/main/kotlin/build/buf/gradle/BufYamlGenerator.kt

```diff
@@ -0,0 +1,47 @@
+package build.buf.gradle
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import org.gradle.api.GradleException
+import java.io.File
+
+class BufYamlGenerator {
+    val yamlMapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+
+    /**
+     * Read the user-supplied buf.yaml file and generate an equivalent file, adding a
+     * `modules` entry for each proto source directory.
+     */
+    fun generate(bufYamlFile: File?, protoDirs: List<String>): String {
+        val bufYaml = bufYamlFile?.let {
+            yamlMapper.readValue(it, object : TypeReference<MutableMap<String, Any>>() {})
+        } ?: mutableMapOf("version" to "v2")
+
+        // Copy existing buf.yaml and force version to v2.
+        val newYaml = bufYaml.toMutableMap().apply {
+            this["version"] = "v2" // Force v2
+        }
```

Comment:
There's a lot of `mutable` being thrown around here, and it's not clear to me what Jackson will return for that TypeReference - there's another call to convert it to a mutable map right afterwards. 

What about a slightly different approach:

```kotlin
// uses `jackson-module-kotlin`'s `readValue` extension
val bufYaml = bufYamlFile?.let { yamlMapper.readValue<Map<String, Any>>(it) }?.orEmpty().toMutableMap()

// now this has first-class indentation
bufYaml["version"] = v2
```

(`newYaml` naming not incorporated here, I didn't want to be so prescripive)

Reply:
Done